plutus-core/executables/pir/Main.hs

```diff
@@ -1,117 +1,242 @@
+               <*> printmode
+
+pPirOptions :: Parser Command
+pPirOptions = hsubparser $
+              command "analyse" (
+                      analyse ("Given a PIR program in flat format, deserialise and analyse the program, " <>
+                               "looking for variables with the largest retained size."))
+           <> command "analyze" (analyse "Same as 'analyse'.")
+           <> command "compile"
+                  (info (Compile <$> pCompileOptions) $
+                   progDesc $
+                   "Given a PIR program in flat format, deserialise it, " <>
+                   "and test if it can be successfully compiled to PLC.")
+           <> command "convert"
+                  (info (Convert <$> pPirConvertOptions)
+                   (progDesc $ "Convert a program between textual and flat-named format."))
+           <> command "optimise" (optimise "Run the PIR optimisation pipeline on the input.")
+           <> command "optimize" (optimise "Same as 'optimise'.")
+           <> command "print"
+                  (info (Print <$> printOpts) $
+                 progDesc $
+                   "Given a PIR program in textual format, " <>
+                   "read it in and print it in the selected format.")
+           where
+             analyse desc = info (Analyse <$> pAnalyseOptions) $ progDesc desc
+             optimise desc = info (Optimise <$> pPirOptimiseOptions) $ progDesc desc
+
+
+---------------- Compilation ----------------
+
+compileToPlc :: Bool -> PirProg () -> Either (PirError UnitProvenance) (PlcTerm UnitProvenance)
+compileToPlc optimise (PIR.Program _ term) = do
     plcTcConfig <- PLC.getDefTypeCheckConfig PIR.noProvenance
-    let pirCtx = defaultCompilationCtx plcTcConfig
-    runExcept $ flip runReaderT pirCtx $ runQuoteT $ PIR.compileTerm pirT
+    let ctx = getCtx plcTcConfig
+    runExcept $ flip runReaderT ctx $ runQuoteT $ PIR.compileTerm term
+  where
+    getCtx :: PLC.TypeCheckConfig PLC.DefaultUni PLC.DefaultFun
+      -> PIR.CompilationCtx PLC.DefaultUni PLC.DefaultFun a
+    getCtx plcTcConfig =
+      PIR.toDefaultCompilationCtx plcTcConfig
+         & PIR.ccOpts . PIR.coOptimize .~ optimise
+    -- See PlutusIR.Compiler.Types.CompilerOpts for other compilation flags,
+    -- including coPedantic, which causes the result of every stage in the
+    -- pipeline to be typechecked.
+
+compileToUplc :: Bool -> PlcProg () -> UplcProg ()
+compileToUplc optimise plcProg =
+    let plcCompilerOpts =
+            if optimise
+            then PLC.defaultCompilationOpts
+            else PLC.defaultCompilationOpts & PLC.coSimplifyOpts . UPLC.soMaxSimplifierIterations .~ 0
+    in flip runReader plcCompilerOpts $ runQuoteT $ PLC.compileProgram plcProg
+
+loadPirAndCompile :: CompileOptions -> IO ()
+loadPirAndCompile (CompileOptions language optimise test inp ifmt outp ofmt mode)  = do
+    pirProg <- readProgram (pirFormatToFormat ifmt) inp
+    when test $ putStrLn "!!! Compiling"
```

Comment:
Perhaps stderr is a better place for this and then you can leave it always enabled (removing the `when test`)

Reply:
I'm not sure about that.  It's probably safer to keep the noise down, but I was trying to preserve the previous behaviour as well.

Reply:
> It should definitely go to stderr, but it looks like we pervasively log to stdout. We should perhaps change that separately.

I've created a ticket for that at PLT-5273.